### DISCUSSION TITLE
Related ids in the Association object design pattern

### DISCUSSION THREAD

**Discussion by UltraDiuve** [sqlalchemy/sqlalchemy#11559] [Usage Questions] [answered]
Dear all,

I have a question related to the [Association object design pattern](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#association-object) and the population of ids on the association object.

I do not understand when the actual ids are populated on the association object, depending on the actions taken on the session. When using the way of creating new association objects like this (from the docs):
```
# create parent, append a child via association
p = Parent()
a = Association(extra_data="some data")
a.child = Child()
p.children.append(a)
```
it looks like a.left_id is None if we do not manually flush.

Let me illustrate with an example from my codebase, where `CommunityOrm` is `Parent`, `CommunityMemberOrm` is `Association` and `CaregiverOrm` is `Child`: 
```python
user_id = UUID('994ca089-0858-4b89-98ce-50159c9d2bc0')
community_id = UUID('eee24c6a-55cb-467c-8607-aad508f657ad')

def join_community(session: Session):
    community = session.execute(
        select(CommunityOrm)
        .where(CommunityOrm.id == community_id)
    ).scalar_one()
    user = session.execute(
        select(CaregiverOrm)
        .where(CaregiverOrm.id == user_id)
    ).scalar_one()
    new_member = CommunityMemberOrm(user=user)
    print("After instantiation of new_member")
    print(f"member user.id: {new_member.user.id}")
    print(f"member user_id: {new_member.user_id}")
    print(f"member community: {new_member.community}")
    print(f"member community_id: {new_member.community_id}")
    print(f"member in session: {new_member in session}")
    community.members.append(new_member)
    print("After appending to community.members")
    print(f"member user.id: {new_member.user.id}")
    print(f"member user_id: {new_member.user_id}")
    print(f"member community.id: {new_member.community.id}")
    print(f"member community_id: {new_member.community_id}")
    print(f"member in session: {new_member in session}")
    session.add(new_member)  # Optional? new_memeber is already in session
    print("After addition of new_member")
    print(f"member user.id: {new_member.user.id}")
    print(f"member user_id: {new_member.user_id}")
    print(f"member community.id: {new_member.community.id}")
    print(f"member community_id: {new_member.community_id}")
    print(f"member in session: {new_member in session}")
    session.flush()
    print("After flush")
    print(f"member user.id: {new_member.user.id}")
    print(f"member user_id: {new_member.user_id}")
    print(f"member community.id: {new_member.community.id}")
    print(f"member community_id: {new_member.community_id}")
    print(f"member in session: {new_member in session}")

with Session(engine, expire_on_commit=False) as session:
    with session.begin():
        join_community(session)
```
This yields:

> After instantiation of new_member
member user.id: 994ca089-0858-4b89-98ce-50159c9d2bc0
member user_id: None
member community: None
member community_id: None
member in session: False
After appending to community.members
member user.id: 994ca089-0858-4b89-98ce-50159c9d2bc0
member user_id: 994ca089-0858-4b89-98ce-50159c9d2bc0
member community.id: eee24c6a-55cb-467c-8607-aad508f657ad
member community_id: None
member in session: True
After addition of new_member
member user.id: 994ca089-0858-4b89-98ce-50159c9d2bc0
member user_id: 994ca089-0858-4b89-98ce-50159c9d2bc0
member community.id: eee24c6a-55cb-467c-8607-aad508f657ad
member community_id: None
member in session: True
After flush
member user.id: 994ca089-0858-4b89-98ce-50159c9d2bc0
member user_id: 994ca089-0858-4b89-98ce-50159c9d2bc0
member community.id: eee24c6a-55cb-467c-8607-aad508f657ad
member community_id: eee24c6a-55cb-467c-8607-aad508f657ad
member in session: True
[/var/folders/l9/9fd5t2p91qv13frdby4713y40000gn/T/ipykernel_98008/2139554265.py:17](https://file+.vscode-resource.vscode-cdn.net/var/folders/l9/9fd5t2p91qv13frdby4713y40000gn/T/ipykernel_98008/2139554265.py:17): SAWarning: Object of type <CommunityMemberOrm> not in session, add operation along 'CaregiverOrm.community_members' will not proceed (This warning originated from the Session 'autoflush' process, which was invoked automatically in response to a user-initiated operation.)
  community.members.append(new_member)

I am surprised by the following:
- `new_member.user_id` is not populated at instantiation: I would have thought that `new_member` being an ORM object, ids related to relationships would be automatically populated as soon as instantiation?
- `new_member.user_id` is populated after appending to community.members. I don't know how this has been triggered? Maybe some kind of autoflush happened, and flush does populate the related ids?
- `new_member.community_id` is not populated after `new_member` has been appended to `community_members` (and this raises issues in our code, as we have another feature that rely on this attribute, to check whether or not a user is a member of a community)
- the last flush triggers the population of `new_member.community_id`

I don't know if this is an expected behavior, but I am confused that one needs to manually flush the session in order for all the "related ids" to be populated. Would using the "related ids" directly from an association object anti-pattern, and one should rather use the "related object . ids"?

See below the orm class definitions, maybe something is off along those lines?
```python
class CommunityOrm(ModelBase):
    __tablename__ = "community"

    id: Mapped[UUID] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(nullable=False)
    description: Mapped[str]

    members: Mapped[List["CommunityMemberOrm"]] = relationship(
        back_populates="community",
        cascade="all, delete-orphan",
    )

class CommunityMemberOrm(ModelBase):
    __tablename__ = "community_member"

    community_id: Mapped[UUID] = mapped_column(
        ForeignKey("community.id", ondelete="CASCADE"),
        nullable=False,
        primary_key=True,
    )
    user_id: Mapped[UUID] = mapped_column(
        ForeignKey("caregiver.id", ondelete="CASCADE"),
        nullable=False,
        primary_key=True,
    )

    community: Mapped[CommunityOrm] = relationship(
        back_populates="members",
    )

    user: Mapped["CaregiverOrm"] = relationship(
        back_populates="community_members",
    )

    def __repr__(self) -> str:
        return (
            f"CommunityMemberOrm("
            f"community_id={self.community_id}, "
            f"user_id={self.user_id})"
        )

class CaregiverOrm(ModelBase):
    __tablename__ = "caregiver"

    id: Mapped[UUID] = mapped_column(
        primary_key=True,
        server_default='gen_random_uuid()',
    )
    uid: Mapped[str]
    user_email: Mapped[str]
    community_members: Mapped[List["CommunityMemberOrm"]] = relationship(
        back_populates="user",
        cascade="all, delete-orphan",
    )
```

Thanks a lot for your insights.

Regards,

Pierre

**Comment by zzzeek** [comment, marked as answer]
Hi -

the ORM works based on the `relationship` construct , by populating attributes or collections with instances of objects, like `myobject.some_collection.append(some_other_object)` or `myobject.other_thing = other_object`.   that produces an internal state that indicates the attribute or collection has a pending change on it.    Nothing happens with database identifiers at this stage.    When the next ORM flush happens, that's when foreign key columns are populated with corresponding primary key values of other tables.     This is necessary for an INSERT scenario since the primary-key-holding object usually doesn't have its own primary key before the INSERT takes place, these are usually server generated.  for the UPDATE scenario, similar behavior is maintained.

There's a FAQ entry that gets into this from the other side a bit at https://docs.sqlalchemy.org/en/20/faq/sessions.html#i-set-the-foo-id-attribute-on-my-instance-to-7-but-the-foo-attribute-is-still-none-shouldn-t-it-have-loaded-foo-with-id-7 but the general behavior of relationships with flush is walked through in the unified tutorial at https://docs.sqlalchemy.org/en/20/tutorial/orm_related_objects.html#cascading-objects-into-the-session

**Reply by UltraDiuve** [reply]
Thanks a lot Mike, this answer is quite clear!
